src/Features/Core/Diagnostics/EngineV1/DiagnosticIncrementalAnalyzer.cs

```diff
@@ -750,34 +821,59 @@ private DiagnosticData UpdatePosition(DiagnosticData diagnostic, SyntaxTree tree
                 helpLink: diagnostic.HelpLink);
         }
 
-        private static IEnumerable<DiagnosticData> GetDiagnosticData(Document document, TextSpan? span, IEnumerable<Diagnostic> diagnostics)
+        private static IEnumerable<DiagnosticData> GetDiagnosticData(Document document, SyntaxTree tree, TextSpan? span, IEnumerable<Diagnostic> diagnostics)
         {
-            return diagnostics != null ? diagnostics.Where(dx => ShouldIncludeDiagnostic(dx, span)).Select(d => DiagnosticData.Create(document, d)) : null;
+            return diagnostics != null ? diagnostics.Where(dx => ShouldIncludeDiagnostic(dx, tree, span)).Select(d => DiagnosticData.Create(document, d)) : null;
```

Comment:
yep, what manish said was what I was thinking too, not other way to solve it I believe. or at least in a sane way. not easy to solve it in v1. so, v2 has new ability to report diagnostics in other document in all actions but not in v1.